bazel/thirdparty/openssl.BUILD

```diff
@@ -0,0 +1,28 @@
+load("@rules_foreign_cc//foreign_cc:defs.bzl", "configure_make")
+
+# TODO
+# - handle --debug vs --release mode flags
```

Comment:
I can help, this can be a select on compilation_mode

Reply:
nice. how can i select on build config, like build:sanitizer? we have some dependencies that opt out of sanitizer or have extra flags that are needed when sanitizers are enabled. pretty sure you can see this in action if you build with --config sanitizers where some unit tests crash

Reply:
For compilation mode switch:
```python
load("@rules_foreign_cc//foreign_cc:defs.bzl", "configure_make")

filegroup(
    name = "all_srcs",
    srcs = glob(["**"]),
)

config_setting(
    name = "debug_mode",
    values = {"compilation_mode": "dbg"},
)

config_setting(
    name = "release_mode",
    values = {"compilation_mode": "opt"},
)

configure_make(
    name = "openssl",
    configure_command = "Configure",
    configure_options = [
        "--libdir=lib",
    ] + select({
        ":debug_mode": ["--debug"],
        ":release_mode": ["--release"],
        "//conditions:default": [],
    }),
    lib_source = ":all_srcs",
    out_shared_libs = [
        "libssl.so",
        "libcrypto.so",
    ],
    visibility = [
        "//visibility:public",
    ],
)
```

Reply:
Otherwise it's recommended to introduce flags to tune things, but there are other ways (features and defines are the old school way). Here's an example of flags for stuff, then you just enable the flags you need in each config: https://github.com/bazelbuild/examples/blob/4d9f5e6c1332ac902a0798b634c67fda4ec5c43b/configurations/select_on_build_setting/BUILD#L10-L13

There are pros/cons of having each build file understand the set of modes/configs that are allowed but IMO it's slightly better to invert that so each build knob is exposed and the different configs can pick and choose all the available build options.

Reply:
> another example is tweaking some stuff in bazel/test.bzl for CI context

As for CI, I recommend defining a flag in bazel/ then having a custom `.bazelrc` for CI that enables whatever build flags you need.

Reply:
> For compilation mode switch:

Fixed

Reply:
>recommend defining a flag in bazel/ then having a custom .bazelrc 

You're talking about something like
```
bool_flag(name = "ci", ...)
config_setting(name = "is_ci", ..)
``
then

`build --//ci=True` or something like that?

Reply:
there are two ways you can do it, let's take openssl as the example:

The compilation mode switch approach (generically) means every library understands the "modes" at in how it's built and those checks will be scattered around the codebase (hopefully there are not many of these but they tend to build up over time).

The other approach is that each library exposes all the knobs and each build mode is just assembling a bunch of knobs. For the openssl case it would look something like:

```
string_flag(name = "mode", values = ["debug", "none", "release"], ...)
config_settings(name = "is_debug", ...)
config_settings(name = "is_release", ...)
```

Then in your .bazelrc each mode does something like:

```
build:debug --@openssl//:mode=debug
build:release --@openssl//:mode=release
build:ci --@openssl//:mode=debug
```

It's more verbose, but then if you want to create a new version you really only need to look in a single spot. Also RE compilation modes, there may be times with openssl you want debug mode on even in `-c fastbuild` (i.e. sanitizers), so decoupling from build modes is nice.

I recommend the latter approach (even though I gave you code for the first approach).

For CI only checks I would recommend a similar approach as CI is really just a special mode (but probably based on sanitizer or release or something?)

Reply:
```
build:debug --@openssl//:mode=debug
build:release --@openssl//:mode=release
build:ci --@openssl//:mode=debug
```

this makes sense. using the compilation mode is cool when i read it in the docs, but you're right, it makes mixing and matching things more confusing compared to this very explicit approach.

fixed.